File: Services/dotnet-services/Core/OmniCure.Core.Infrastructure/Persistence/Repositories/QueryExecutor.cs
using Microsoft.EntityFrameworkCore;
using OmniCure.Core.Domain.Interfaces.Persistence;
using OmniCure.Core.Shared.Pagination;

namespace OmniCure.Core.Infrastructure.Persistence.Repositories;

public class QueryExecutor : IQueryExecutor
{
    public async Task<PagedList<TResult>> ToPagedListAsync<TResult>(
        IQueryable<TResult> query,
        int pageNumber,
        int pageSize,
        CancellationToken cancellationToken = default)
    {
        var totalCount = await query.CountAsync(cancellationToken);

        var items = await query
            .Skip((pageNumber - 1) * pageSize)
            .Take(pageSize)
            .ToListAsync(cancellationToken);

        return new PagedList<TResult>(items, totalCount, pageNumber, pageSize);
    }
}

File: Services/java-services/Core/CoreInfra/src/main/java/ai/omnicure/core/infra/repository/QueryExecutor.java
package ai.omnicure.core.infra.repository;

import ai.omnicure.core.domain.repository.IQueryExecutor;
import ai.omnicure.core.shared.pagination.PagedList;
import org.springframework.data.domain.Page;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;

@Component
public class QueryExecutor implements IQueryExecutor {

    @Override
    public <T> PagedList<T> toPagedList(Iterable<T> items, long totalCount, int pageNumber, int pageSize) {
        List<T> list = new ArrayList<>();
        items.forEach(list::add);
        return new PagedList<>(list, totalCount, pageNumber, pageSize);
    }

    public <T> PagedList<T> fromPage(Page<T> page) {
        return new PagedList<>(
            page.getContent(),
            page.getTotalElements(),
            page.getNumber() + 1,
            page.getSize()
        );
    }
}
